src/metabase/api/search.clj

```diff
@@ -380,17 +308,14 @@
                                      query)}
           _            (log/tracef "Searching with query:\n%s" (u/pprint-to-str search-query))
           results      (db/query search-query :max-rows search-max-results)]
-      ;; sort by [score model name]
-      (sort-by (juxt
-                (comp - (partial search/score (:search-string search-ctx)))
-                (comp model->sort-position :model)
-                :name)
-               (for [row results
-                     :when (check-permissions-for-model row)]
-                 ;; MySQL returns `:favorite` and `:archived` as `1` or `0` so convert those to boolean as needed
-                 (-> row
-                     (update :favorite bit->boolean)
-                     (update :archived bit->boolean)))))))
+      (scoring/sort-results
```

Comment:
i know you limited the results to 1000, but i think it would be nice if this would use a transducer on a reducible collection rather than realizing the entire result set first.

Reply:
Won't sorting inherently need to realize the entire set? I could rewrite the `for` expression with transducers but am not sure that'd gain us anything

Footnote: watched Rich's talk introducing transducers live and in person, but have never gotten to use them in the wild, so I'm probably a little ignorant here.

Reply:
sorting yes. but the scoring and discarding of the 0 scored items can be done row per row out of the db. I'm in the middle of a 150 file rename refactor so i can't check out the branch right now but somewhere you do the query. and at that spot you can make a transducer that scores and then removes the 0 scored items. For an example you can look at https://github.com/metabase/metabase/blob/master/src/metabase/sync/analyze/fingerprint.clj#L66 where a transducer is passed into the qp (called `rff`) and https://github.com/metabase/metabase/blob/master/src/metabase/db/metadata_queries.clj#L144 where it is used

Reply:
👍 on transducing everything... the code is a little harder to write but the memory usage will be improved dramatically. Do we really want to return **1000** results? That seems like a lot. We're not going to show 1000 results on screen at once

Reply:
1000 results and then sort and take N is probably best. Since we score on the clojure side we need to taken enough to ensure that we don't get 500 results with a score of 0? Or perhaps 500 results all of score 1.0 and miss the score 3.0 results? I'm a bit worried about this particularly as we start looking at card sql queries. We might get 1000 cards matching on a table name in the sql and not get "Tablename Dashboard" as a card title

Reply:
Just changed this to use a transducer.

I'm going to hold off on tinkering the max-results number (1000) until we

a) know what other columns could result in a score of 0

b) make decisions around pagination

Reply:
and yeah, it should be two numbers: the `LIMIT n` on the DB, and the `(take m)` in the xf.

`m` is a Kyle question, and `n` will have to be arrived at experimentally based on performance and the percentage of results thrown out

Reply:
your transducer needs to go into the query processor. `(db/query search-query :max-rows search-config/search-max-results)` will fetch all the results and then work on them after. we want to get a cursor into the results and as the rows are realized one by one, score them, and discard them if they are 0. This allows our final realized collection to be only positive scoring results which are then sorted and a subset taken.

Reply:
```
(transduce (comp (map (fn [{:keys [name]}] [name (count name)]))
                 (filter (fn [[name length]] (> length 14))))
           conj
           []
           (db/reducible-query {:select [:name]
                                :from [:metabase_field]}))
[["MUNICIPALITY-ID" 15]]
```

toucan exposes `reducible-query` that will return a reducible collection which we can transduce over. Here i'm getting all the field names in the query. And row by row, getting the length of the name and filtering where the name is longer than 14. My reducing function is `conj` which just sticks the results of this in a vector.

So I think yours needs to be something similar to this except:
- rather than counting the length of the name stick a score on it
- filter to the score threshold
- conj will be the same most likely.
Once you do this you can sort and then take. But this way the realized collection isn't all the results that come back but only the results with positive scores. Keeps memory pressure down.

Reply:
@dpsutton thanks for the major assist! Ready for re-review